test/invariants/symbolic/Protocol.t.sol

```diff
@@ -0,0 +1,302 @@
+import {BToken} from 'contracts/BToken.sol';
+
+contract HalmosBalancer is HalmosTest {
+  // System under test
+  BCoWFactory factory;
+  BConst bconst;
+  BMath bmath;
+  BNum_exposed bnum;
+
+  address solutionSettler;
+  bytes32 appData;
+
+  FuzzERC20[] tokens;
+  BCoWPool pool;
+
+  address currentCaller = svm.createAddress('currentCaller');
+  // address currentCaller = address(234);
+
+  constructor() {
+    solutionSettler = address(new MockSettler());
+    factory = new BCoWFactory(solutionSettler, appData);
+    bconst = new BConst();
+    bmath = new BMath();
+    bnum = new BNum_exposed();
+    pool = BCoWPool(address(factory.newBPool()));
+
+    // max bound token is 8
+    for (uint256 i; i < 5; i++) {
+      FuzzERC20 _token = new FuzzERC20();
+      _token.initialize('', '', 18);
+      tokens.push(_token);
+
+      _token.mint(address(this), 10 ether);
+      _token.approve(address(pool), 10 ether);
+
+      uint256 _poolWeight = bconst.MAX_WEIGHT() / 5;
+
+      pool.bind(address(_token), 10 ether, _poolWeight);
+    }
+
+    pool.finalize();
+  }
+
+  /// @custom:property-id 0
+  /// @custom:property BFactory should always be able to deploy new pools
+  function check_deploy() public {
+    assert(factory.SOLUTION_SETTLER() == solutionSettler);
+    assert(pool.isFinalized());
+  }
+
+  /// @custom:property-id 1
+  /// @custom:property BFactory should always be able to deploy new pools
+  function check_BFactoryAlwaysDeploy(address _caller) public {
+    // Precondition
+    vm.assume(_caller != address(0));
+    vm.prank(_caller);
+
+    // Action
+    try factory.newBPool() returns (IBPool _newPool) {
+      // Postcondition
+      assert(address(_newPool).code.length > 0);
+      assert(factory.isBPool(address(_newPool)));
+      assert(!_newPool.isFinalized());
+    } catch {
+      assert(false);
+    }
+  }
+
+  /// @custom:property-id 2
+  /// @custom:property BFactory's blab should always be modifiable by the current blabs
+  function check_blabAlwaysModByBLab() public {
+    // Precondition
+    address _currentBLab = factory.getBLabs();
+
+    vm.prank(currentCaller);
+
+    // Action
+    try factory.setBLabs(address(123)) {
+      // Postcondition
+      assert(_currentBLab == currentCaller);
+    } catch {
+      assert(_currentBLab != currentCaller);
+    }
+  }
+
+  /// @custom:property-id 3
+  /// @custom:property BFactory should always be able to transfer the BToken to the blab, if called by it
+  function check_alwaysCollect() public {
+    // Precondition
+    address _currentBLab = factory.getBLabs();
+
+    vm.prank(currentCaller);
+
+    // Action
+    try factory.collect(pool) {
+      // Postcondition
+      assert(_currentBLab == currentCaller);
+    } catch {
+      assert(_currentBLab != currentCaller);
+    }
+  }
+  /// @custom:property-id 7
+  /// @custom:property total weight can be up to 50e18
+  /// @dev Only 2 tokens are used, to avoid hitting the limit in loop unrolling
```

Comment:
not finding it clear, why not 3 tokens possible?

Reply:
it's possible, just need to pass an extra-arg to halmos (to support loop which are one iteration longer), but writing a first version of all tests was prioritised tho